File: anthropic_sdk_steps/step09_review_skill.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25
//DEPS com.anthropic:anthropic-java:2.27.0
//DEPS org.slf4j:slf4j-nop:1.7.36
//SOURCES InnerHarness.java

import com.anthropic.client.okhttp.AnthropicOkHttpClient;
import com.anthropic.models.beta.messages.*;
import com.anthropic.models.messages.Model;
import com.fasterxml.jackson.annotation.JsonClassDescription;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.List;

// Step 9: Review skill — feedback, inferential. LLM-as-judge.
// "Instructions how to review, feedback, inferential, e.g. Skills." — Böckeler
// "Separating the agent doing the work from the agent judging it proves
//  to be a strong lever." — Anthropic

static final int MAX_LINT_ITERATIONS = 5;

void main(String[] args) {
    IO.println("=== Step 9: Review Skill (Feedback, Inferential) ===");

    var agentsMd = readFile("outer-harness/AGENTS.md");
    var agent = new InnerHarness.Agent(25);
    agent.prependSystemPrompt("# Project Conventions\n\n" + agentsMd);

    agent.registerTool(LoadSkill.class, tu -> tu.input(LoadSkill.class).execute());
    agent.registerTool(RunCodemod.class, tu -> tu.input(RunCodemod.class).execute());

    // Linter + ArchUnit sensor (same as Step 8)
    var iterations = new AtomicInteger(0);
    agent.afterWriteHook = (filePath, content) -> {
        if (!filePath.endsWith(".java")) return null;
        if (iterations.get() >= MAX_LINT_ITERATIONS) return null;
        iterations.incrementAndGet();

        var feedback = new StringBuilder();

        try {
            var proc = new ProcessBuilder("jbang", "outer-harness/linters/ResultUnwrapChecker.java", filePath)
                    .redirectErrorStream(true).start();
            var output = new String(proc.getInputStream().readAllBytes());
            if (proc.waitFor() != 0) feedback.append("LINTER FEEDBACK:\n").append(output).append("\n\n");
        } catch (Exception e) { /* skip */ }

        if (filePath.contains("bookshelf/src/main")) {
            try {
                var proc = new ProcessBuilder("bash", "-c", "cd bookshelf && mvn test -Dtest=BookshelfArchitectureTest -q 2>&1")
                        .redirectErrorStream(true).start();
                var output = new String(proc.getInputStream().readAllBytes());
                if (proc.waitFor() != 0) feedback.append("ARCHUNIT FEEDBACK:\n").append(output).append("\n\n");
            } catch (Exception e) { /* skip */ }
        }

        if (feedback.isEmpty()) { iterations.set(0); return null; }
        return "MANDATORY FIX REQUIRED:\n\n" + feedback
                + "You MUST immediately fix these issues. Do NOT explain the problem, do NOT ask the user. "
                + "Move or rewrite the offending code to comply, then call write_file now.";
    };

    // THE KEY: afterDoneHook runs the LLM-as-judge review
    var reviewDone = new AtomicBoolean(false);
    agent.afterDoneHook = (paramsBuilder) -> {
        if (reviewDone.get()) return false; // Only review once
        reviewDone.set(true);

        IO.println("\n[review] Running LLM-as-judge code review...");

        // Get the diff
        String diff;
        try {
            var proc = new ProcessBuilder("git", "diff").redirectErrorStream(true).start();
            diff = new String(proc.getInputStream().readAllBytes());
            proc.waitFor();
        } catch (Exception e) {
            IO.println("[review] Could not get diff: " + e.getMessage());
            return false;
        }

        if (diff.isBlank()) {
            IO.println("[review] No changes to review.");
            return false;
        }

        // Load the review skill
        var reviewPrompt = readFile("outer-harness/skills/code-review.md");

        // Make a separate API call — the sub-agent pattern
        var reviewClient = AnthropicOkHttpClient.fromEnv();
        var reviewParams = MessageCreateParams.builder()
                .model(Model.CLAUDE_HAIKU_4_5)
                .maxTokens(4096)
                .system(reviewPrompt)
                .addUserMessage("Review this diff against the bookshelf codebase conventions:\n\n```diff\n" + diff + "\n```")
                .build();

        var reviewMessage = reviewClient.beta().messages().create(reviewParams);
        var reviewText = new StringBuilder();
        for (var block : reviewMessage.content()) {
            block.text().ifPresent(t -> reviewText.append(t.text()));
        }

        var review = reviewText.toString();
        IO.println("[review] " + review);

        // If review found issues, inject them back
        if (review.toLowerCase().contains("[critical]") || review.toLowerCase().contains("[high]")) {
            IO.println("[review] Issues found — injecting feedback for agent to fix.");
            paramsBuilder.addUserMessage(
                    "A code reviewer found the following issues with your changes. Please fix them:\n\n" + review);
            return true; // Continue the agent loop
        }

        IO.println("[review] Review passed.");
        return false;
    };

    var task = args.length > 0 ? String.join(" ", args)
            : "Add a test for BookService.findAllBooks that verifies it returns both books when two are added. Write the test data inline (don't use fixtures).";
    // NOTE: Deliberately asks to inline test data, which the review skill catches.

    agent.run(task);
}

@JsonClassDescription("Run the result-type-rewrite codemod to convert a method from throwing exceptions to returning Result<T, BookshelfError>. Much more reliable than manual editing for this transformation.")
static class RunCodemod {
    @JsonPropertyDescription("Path to the Java file to modify")
    public String file;
    @JsonPropertyDescription("Name of the method to rewrite")
    public String method;

    public String execute() {
        try {
            var proc = new ProcessBuilder("jbang", "outer-harness/codemods/ResultTypeRewrite.java", file, method)
                    .redirectErrorStream(true).start();
            var output = new String(proc.getInputStream().readAllBytes());
            var exitCode = proc.waitFor();
            return output + "\n[exit code: " + exitCode + "]";
        } catch (Exception e) {
            return "Error running codemod: " + e.getMessage();
        }
    }
}

@JsonClassDescription("Load a skill by name for detailed guidance. Available: how-to-write-tests, code-review, architecture-review")
static class LoadSkill {
    @JsonPropertyDescription("The skill name")
    public String name;

    public String execute() {
        try { return Files.readString(Path.of("outer-harness/skills/" + name + ".md")); }
        catch (IOException e) { return "Skill not found: " + name; }
    }
}

String readFile(String path) {
    try { return Files.readString(Path.of(path)); }
    catch (IOException e) { return ""; }
}

File: python/steps/step09_review_skill.py
#!/usr/bin/env python3
"""Step 9: Review skill -- feedback, inferential. LLM-as-judge.
'Instructions how to review, feedback, inferential, e.g. Skills.' -- Bockeler"""

import subprocess
import sys

from inner_harness import Agent, string_property

MAX_LINT_ITERATIONS = 5


def load_skill(arguments):
    name = arguments.get("name", "")
    if not name:
        return "Error: name is required."
    try:
        with open(f"outer_harness/skills/{name}.md") as f:
            return f.read()
    except FileNotFoundError:
        return f"Skill not found: {name}"


def run_codemod(arguments):
    file = arguments.get("file", "")
    method = arguments.get("method", "")
    if not file:
        return "Error: file is required."
    if not method:
        return "Error: method is required."
    try:
        result = subprocess.run(
            ["python", "outer_harness/codemods/result_type_rewrite.py", file, method],
            capture_output=True, text=True, timeout=60,
        )
        return result.stdout + result.stderr + f"\n[exit code: {result.returncode}]"
    except Exception as e:
        return f"Error running codemod: {e}"


def main():
    print("=== Step 9: Review Skill (Feedback, Inferential, Ollama) ===")

    # Load AGENTS.md
    try:
        with open("outer_harness/AGENTS.md") as f:
            agents_md = f.read()
    except OSError:
        agents_md = ""

    agent = Agent(max_turns=25)
    agent.prepend_system_prompt("# Project Conventions\n\n" + agents_md)

    # Register skill and codemod tools
    agent.register_tool(
        "load_skill",
        "Load a skill by name for detailed guidance. "
        "Available: how-to-write-tests, code-review, architecture-review",
        {"name": string_property("The skill name")},
        load_skill,
    )
    agent.register_tool(
        "run_codemod",
        "Run the result-type-rewrite codemod to convert a method from throwing "
        "exceptions to returning Result. Much more reliable than manual editing "
        "for this transformation.",
        {"file": string_property("Path to the Python file to modify"),
         "method": string_property("Name of the method to rewrite")},
        run_codemod,
    )

    # Linter + import checker sensor (same as Step 8)
    lint_iterations = [0]

    def combined_sensor_hook(file_path, content):
        if not file_path.endswith(".py"):
            return None
        if lint_iterations[0] >= MAX_LINT_ITERATIONS:
            return None
        lint_iterations[0] += 1

        feedback = []

        try:
            proc = subprocess.run(
                ["python", "outer_harness/linters/result_unwrap_checker.py", file_path],
                capture_output=True, text=True, timeout=30,
            )
            if proc.returncode != 0:
                feedback.append("LINTER FEEDBACK:\n" + proc.stdout + proc.stderr)
        except Exception:
            pass

        if "bookshelf/" in file_path and "bookshelf/tests/" not in file_path:
            try:
                proc = subprocess.run(
                    ["python", "outer_harness/arch/import_checker.py", "bookshelf"],
                    capture_output=True, text=True, timeout=30,
                )
                if proc.returncode != 0:
                    feedback.append("IMPORT CHECK FEEDBACK:\n" + proc.stdout + proc.stderr)
            except Exception:
                pass

        if not feedback:
            lint_iterations[0] = 0
            return None

        return (
            "MANDATORY FIX REQUIRED:\n\n"
            + "\n\n".join(feedback)
            + "\n\nYou MUST immediately fix these issues. Do NOT explain the problem, "
            "do NOT ask the user. "
            "Move or rewrite the offending code to comply, then call write_file now."
        )

    agent.after_write_hook = combined_sensor_hook

    # THE KEY: afterDoneHook runs the LLM-as-judge review
    review_done = [False]  # mutable flag in closure

    def review_hook(running_agent):
        if review_done[0]:
            return False
        review_done[0] = True

        print("\n[review] Running local Ollama code review...")

        # Get git diff
        try:
            proc = subprocess.run(
                ["git", "diff"],
                capture_output=True, text=True, timeout=30,
            )
            diff = proc.stdout
        except Exception as e:
            print(f"[review] Could not get diff: {e}")
            return False

        if not diff.strip():
            print("[review] No changes to review.")
            return False

        # Read review skill
        try:
            with open("outer_harness/skills/architecture-review.md") as f:
                review_prompt = f.read()
        except OSError:
            review_prompt = "Review the following diff for code quality issues."

        review = running_agent.complete(
            review_prompt,
            "Review this diff:\n\n```diff\n" + diff + "\n```",
        )

        print(f"[review] {review}")

        if "[critical]" in review.lower() or "[high]" in review.lower():
            print("[review] Issues found - injecting feedback for agent to fix.")
            running_agent.add_user_message(
                "A code reviewer found the following issues with your changes. "
                "Please fix them:\n\n" + review
            )
            return True

        print("[review] Architecture review passed.")
        return False

    agent.after_done_hook = review_hook

    task = (
        " ".join(sys.argv[1:])
        if len(sys.argv) > 1
        else "Add a test for BookService.find_all_books that verifies it returns both "
             "books when two are added. Write the test data inline (don't use fixtures)."
    )

    agent.run(task)


if __name__ == "__main__":
    main()
